### PR TITLE
feat(hard_fork): part2 (Delyayed ack callback)

### PR DESCRIPTION
**Pull request by mtsitrin**
* remove `Reverted` status
* handle rollapp packet deletion in case of `HardFork`
* removed `x/delayedack` invariants as it's broken (since lazy finalization)

### CODE REVIEW

File: x/delayedack/keeper/fraud.go [hunk 1]
```diff
@@ -28,22 +25,25 @@ func (k Keeper) HandleFraud(ctx sdk.Context, rollappID string, ibc porttypes.IBC
...
-		}
-		// Update status to reverted
-		_, err := k.UpdateRollappPacketWithStatus(ctx, rollappPacket, commontypes.Status_REVERTED)
-		if err != nil {
-			logger.Error("error reverting IBC rollapp packet", append(logContext, "error", err.Error())...)
-			return err
+		} else {
+			// for incoming packets, we need to reset the packet receipt
+			ibcPacket := rollappPacket.Packet
+			k.channelKeeper.SetPacketReceipt(ctx, ibcPacket.GetDestPort(), ibcPacket.GetDestChannel(), ibcPacket.GetSequence())
```
Comment (omritoptix) [inline x/delayedack/keeper/fraud.go:38]: I don't see how this is deleting the packet reciept? 

from the ADR

```go
	// remove receipt
    ibcPacket := rollappPacket.Packet
    receiptKey := host.PacketReceiptKey(ibcPacket.GetDestPort(), ibcPacket.GetDestChannel(), ibcPacket.GetSequence())
    channelKeeperStore.Delete(receiptKey)
```

File: x/delayedack/keeper/rollapp_packet.go [hunk 2]
```diff
@@ -178,3 +181,24 @@ func (k Keeper) deleteRollappPacket(ctx sdk.Context, rollappPacket *commontypes.
...
+
+func (k Keeper) GetPendingPacketsUntilLatestHeight(ctx sdk.Context, rollappID string) ([]commontypes.RollappPacket, uint64, error) {
+	// Get rollapp's latest finalized height
+	latestFinalizedHeight, err := k.GetRollappLatestFinalizedHeight(ctx, rollappID)
+	if err != nil {
+		return nil, 0, fmt.Errorf("get latest finalized height: rollapp '%s': %w", rollappID, err)
+	}
+
+	// Get all pending rollapp packets until the latest finalized height
+	return k.ListRollappPackets(ctx, types.PendingByRollappIDByMaxHeight(rollappID, latestFinalizedHeight)), latestFinalizedHeight, nil
```
[the remark this answers is not in the document]
Comment (mtsitrin) [reply x/delayedack/keeper/rollapp_packet.go:193]: this code haven't changed. copied from `x/delayedack/keeper/finalize.go`
it's not used for fraud, it used by the query `GetPendingPacketsByReceiver`
I'll rename for clarity

File: x/delayedack/keeper/fraud.go (hunk 1 above)
Comment (mtsitrin) [reply x/delayedack/keeper/fraud.go:38]: u right!
will fix

File: x/delayedack/keeper/fraud.go [hunk 3]
```diff
@@ -9,14 +9,11 @@ import (
...
-	rollappPendingPackets := k.ListRollappPackets(ctx, types.ByRollappIDByStatus(rollappID, commontypes.Status_PENDING))
-	if len(rollappPendingPackets) == 0 {
-		return nil
-	}
+func (k Keeper) HandleHardFork(ctx sdk.Context, rollappID string, height uint64, ibc porttypes.IBCModule) error {
 	logger := ctx.Logger().With("module", "DelayedAckMiddleware")
-	logger.Info("reverting IBC rollapp packets", "rollappID", rollappID)
+
+	// Get all the pending packets
+	rollappPendingPackets := k.ListRollappPackets(ctx, types.PendingByRollappIDFromHeight(rollappID, height))
```
[the remark this answers is not in the document]
Comment (mtsitrin) [reply x/delayedack/keeper/fraud.go:16]: pls open research topic

File: x/delayedack/keeper/fraud.go [hunk 4]
```diff
@@ -28,22 +26,31 @@ func (k Keeper) HandleFraud(ctx sdk.Context, rollappID string, ibc porttypes.IBC
...
 		}
 
+		// refund all pending outgoing packets
 		if rollappPacket.Type == commontypes.RollappPacket_ON_ACK || rollappPacket.Type == commontypes.RollappPacket_ON_TIMEOUT {
-			// refund all pending outgoing packets
 			// we don't have access directly to `refundPacketToken` function, so we'll use the `OnTimeoutPacket` function
 			err := ibc.OnTimeoutPacket(ctx, *rollappPacket.Packet, rollappPacket.Relayer)
 			if err != nil {
 				logger.Error("failed to refund reverted packet", append(logContext, "error", err.Error())...)
 			}
```
Comment (omritoptix) [inline x/delayedack/keeper/fraud.go:35]: should be changed to follow the ADR:
1. timeout - only deleting the rollapp packet
2. on_ack - to delete packet and re-create the commitment

**Pull request merged**